server/src/storage.rs

```diff
@@ -1,10 +1,43 @@
 use super::{ReportTimestamp, SignedReport};
-use eyre::ErrReport;
+use eyre::{eyre, ErrReport};
+use rand::seq::SliceRandom;
+use rand::rngs::OsRng;
+use std::collections::HashMap;
 
-pub(crate) fn save(report: SignedReport) -> Result<(), ErrReport> {
-    unimplemented!();
+#[derive(Default)]
+struct StorageEntry {
+    reports: Vec<SignedReport>,
+    is_randomized: bool,
 }
 
-pub(crate) fn get(timeframe: ReportTimestamp) -> Result<Vec<u8>, ErrReport> {
-    unimplemented!();
+#[derive(Default)]
+pub struct Storage {
+    map: HashMap<ReportTimestamp, StorageEntry>
+}
+
+impl Storage {
+    pub(crate) fn save(&mut self, report: SignedReport) -> Result<(), ErrReport> {
+        let now = ReportTimestamp::now()?;
+        let entry = self.map.entry(now).or_default();
+        if entry.is_randomized {
+            return Err(eyre!("Attempted to save for entry that has been read from. Is time broken?"))
+        }
+        entry.reports.push(report);
+        Ok(())
+    }
+
+    pub(crate) fn get(&mut self, timeframe: ReportTimestamp) -> Result<Vec<u8>, ErrReport> {
+        let entry = self.map.get_mut(&timeframe).ok_or(eyre!("No entries for this timestamp"))?;
+        if !entry.is_randomized {
+            entry.is_randomized = true;
+            entry.reports.shuffle(&mut OsRng)
+        }
+
+        let mut bytes: Vec<u8> = vec![];
```

Comment:
We might want to just cache this on the struct

Reply:
One way that could work would be to change StorageEntry to an enum of `Open(Vec<SignedReport>)` and `Closed(Bytes)`. This way there's no need to keep the `is_randomized` separately.

Reply:
I'm already doing something like that, though I'm still holding on to the original vec